### PR TITLE
Tree widget: integrate `"@itwin/presentation-hierarchies-react"`

### PR DESCRIPTION
**Pull request by saskliutas**
Initial implementation bringing `ModelsTree` component based on the new `"@itwin/presentation-hierarchies-react"` library.

### CODE REVIEW

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import { VisibilityStatus } from "../../VisibilityTreeEventHandler";
+
+interface HierarchyVisibilityHandler {
+  getVisibilityStatus: (node: HierarchyNode) => Promise<VisibilityStatus>;
+  changeVisibility: (node: HierarchyNode, on: boolean) => Promise<void>;
+  onVisibilityChange: BeEvent<() => void>;
+  dispose: () => void;
+}
+
+interface UseNodesVisibilityProps {
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts:20]: should be `UseHierarchyVisibilityProps`?

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+          hierarchyLevelDetails.getInstanceKeysIterator({
+            instanceFilter,
+            hierarchyLevelSizeLimit: hierarchyLevelDetails.sizeLimit ?? defaultHierarchyLevelSizeLimit,
+          }),
+        );
+        return `Current filter matching instances count: ${instanceKeys.length}`;
+      } catch (e) {
+        if (e instanceof RowsLimitExceededError) {
+          return `Current filter exceeds instances count of ${e.limit}`;
+        }
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx:143]: I think here we should either rethrow the error or return a message saying there was an error getting the instances count.

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+interface UseHierarchyFilteringOwnProps {
+  imodel: IModelConnection;
+  defaultHierarchyLevelSizeLimit: number;
+}
+
+type UseHierarchyFilteringProps = UseHierarchyFilteringOwnProps & Pick<UseTreeResult, "getHierarchyLevelDetails">;
+
+/** @internal */
+export function useHierarchyFiltering({ imodel, defaultHierarchyLevelSizeLimit, getHierarchyLevelDetails }: UseHierarchyFilteringProps) {
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx:28]: can we change the name of this to `useHierarchyLevelFiltering`?

File: packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsVisibilityHandler.ts [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+      [],
+    );
+  }
+
+  private getAssemblyElementIds(assemblyId: Id64String) {
+    return this._elementIdsCache.getAssemblyElementIds(assemblyId);
+  }
+
+  // istanbul ignore next
+  private getGroupedElementIds(node: GroupingHierarchyNode & { key: GroupingNodeKey }) {
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsVisibilityHandler.ts:362]: is the ` & { key: GroupingNodeKey }` part needed?

**Review by grigasp — COMMENTED** [review]
I think at this point we can change the term "experimental" to "stateless".

File: packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsTreeComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  extends Pick<ExperimentalModelsTreeProps, "getSchemaContext" | "density" | "hierarchyLevelConfig" | "selectionMode"> {
+  headerButtons?: Array<(props: ModelsTreeHeaderButtonProps) => React.ReactNode>;
+  selectionStorage: SelectionStorage;
+}
+
+/**
+ * A component that renders [[ModelsTree]] and a header with filtering capabilities
+ * and header buttons.
+ * @public
+ */
```
Comment (jasdom) [inline packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsTreeComponent.tsx:39]:
```suggestion
/**
 * A component that renders [[StatelessModelsTree]] and a header with filtering capabilities
 * and header buttons.
 * @beta
 */
```

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    const changeVisibility = (node: PresentationHierarchyNode, checked: boolean) => {
+      void handler.changeVisibility(node.nodeData, checked);
+      const status = statusMap.current.get(node.id);
+      if (!status) {
+        return;
+      }
+      status.status.state = checked ? "visible" : "hidden";
+      status.status.tooltip = undefined;
+      setState((prev) => ({ ...prev, getCheckboxStatus: createStatusGetter(statusMap, (node) => calculate.next(node)) }));
```
Comment (Yato333) [inline packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts:67]: Why do you update this state with the same value in multiple places?

File: packages/itwin/tree-widget/src/components/trees/experimental/common/FocusedInstancesContextProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  imodelKey,
+  children,
+}: PropsWithChildren<{ selectionStorage: SelectionStorage; imodelKey: string }>) {
+  const [state, setState] = useState<FocusedInstancesContext>({
+    enabled: false,
+    toggle: () => {
+      setState((prev) => ({ ...prev, enabled: !prev.enabled }));
+    },
+  });
+  const enabled = state.enabled;
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/FocusedInstancesContextProvider.tsx:24]: I wanted to make it clear that `useEffect` depends only on `enabled` values from state.

File: packages/itwin/tree-widget/src/components/trees/experimental/common/components/VisibilityTree.tsx [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    reloadTree,
+    selectNodes,
+    setFormatter: _,
+    ...treeProps
+  } = useUnifiedSelectionTree({
+    imodelAccess,
+    getHierarchyDefinition,
+    getFilteredPaths,
+    imodelKey: imodel.key,
+    sourceName: "ExperimentalModelsTree",
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/components/VisibilityTree.tsx]: Yes. Now it is passed as a prop.

File: packages/itwin/tree-widget/src/components/trees/experimental/common/components/VisibilityTree.tsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+  if (rootNodes === undefined) {
+    return (
+      <Flex alignItems="center" justifyContent="center" flexDirection="column" style={{ width, height }}>
+        <Delayed show={true}>
+          <ProgressRadial size="large" />
+        </Delayed>
+      </Flex>
+    );
+  }
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/components/VisibilityTree.tsx:110]: I don't like how overlay looks when there are no nodes. So if we do not have any nodes I think spinner looks better.

File: packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsTree.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  imodel: IModelConnection;
+  height: number;
+  width: number;
+  activeView: Viewport;
+  getSchemaContext: (imodel: IModelConnection) => SchemaContext;
+  filter: string;
+  density?: "default" | "enlarged";
+  hierarchyLevelConfig?: Omit<HierarchyLevelConfig, "isFilteringEnabled">;
+  selectionMode?: SelectionMode;
+}
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsTree.tsx]: Defined `StatelessModelsTreeProps` using `VisibilityTreeProps`

File: packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsTreeComponent.tsx (hunk 5 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsTreeComponent.tsx]: Updated comment.

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts (hunk 6 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts:67]: Updating state causes all components that use `getCheckboxStatus` to rerender.

File: apps/test-viewer/src/UiProvidersConfig.tsx [hunk 11]
```diff
@@ -130,10 +154,23 @@ const configuredUiItems = new Map<string, UiItem>([
                   />
                 ),
               },
+              {
+                id: `experimental-${ModelsTreeComponent.id}`,
+                getLabel: () => `${ModelsTreeComponent.getLabel()} (Experimental)`,
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply apps/test-viewer/src/UiProvidersConfig.tsx]: Updated

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts (hunk 1 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyVisibility.ts]: Renamed

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx (hunk 2 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx:143]: Updated it to return error message

File: packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx (hunk 3 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx]: Renamed

File: packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsVisibilityHandler.ts (hunk 4 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsVisibilityHandler.ts]: Removed

**Review by grigasp — APPROVED** [review]

**Review by Yato333 — APPROVED** [review]

**Pull request merged**